option java_package = "com.nextev.cvs_proto.protobuf";
option java_outer_classname = "NextEVMessage";

// current version is 19

message Message {
    optional int32 version = 1; // message protocol version
    optional string id = 2; // message id
    optional int64 publish_ts = 3; // message publish time(server time)
    optional int64 ttl = 4; // message expiration (second)

    optional MessageType type = 5; // 2-level message type
    optional string sub_type = 6;

    repeated ParamType params = 7; //generic key-value parameters
    optional QoS qos = 8 [default = LEAST_ONCE]; // same as mqtt qos

    optional int32 target = 9;//target ecu  SA:1 M4:2 A35:3 CDC:4 ADC:5

    optional PlatformType platform_type = 10;

    enum MessageType {
        NOTIFICATION = 0;
        CONTROL_COMMAND = 1;
        COMMAND_RESULT = 2;
        DATA_REPORT = 3;
        CLIENT_STATUS = 4;
    }

    enum QoS {
        MOST_ONCE = 0;
        LEAST_ONCE = 1;
        EXACTLY_ONCE = 2;
    }

    message ParamType {
        optional string key = 1;
        optional bytes value = 2;
    }

    enum PlatformType {
        NT1 = 0;
        NT2 = 1;
        NT1_2 = 2; // NT1.2
    }

}